<div style="display: flex; align-items: center; gap: 20px; font-family: Arial, sans-serif;">
    <div style="margin-right: 30px;">
        <img src="../Sources/Logo.png" width="300" style="border-radius: 10px;"/>
    </div>
    <div style="line-height: 1.5;">
        <strong style="font-size: 18px; display: block; margin-bottom: 5px;">
            <a href='https://t.me/M4m4ev_Alex' style="text-decoration: none; color: #1a73e8;">
                Репетитор по информатике | Мамаев Алексей
            </a>
        </strong>
        <span style="font-size: 16px; color: #333;">
            Разбор <a href='https://euler.jakumo.org/problem/9-osobaia-troika-pifagora' style="text-decoration: none; color: #1a73e8;">
            9-й задачи</a> проекта Эйлера
        </span>
        <span style="font-size: 16px; color: #333;">
            <br><br><br><br><br><br>ТГК: <a href='https://t.me/M4m4ev_Alex' style="text-decoration: none; color: #1a73e8;">@M4m4ev_Alex</a><br>
        </span>
    </div>
</div>

## Задача 9

Тройкой Пифагора называются три натуральных числа `a < b < c`, для которых выполняется равенство

$$
a^2 + b^2 = c^2
$$

Например:

$$
3^2 + 4^2 = 9 + 16 = 25 = 5^2
$$

Известно, что существует единственная тройка Пифагора, для которой

$$
a + b + c = 1000
$$

Необходимо найти произведение

$$
abc
$$

Ответ: 31875000

## Используемые библиотеки

В данном разборе используются стандартные библиотеки Python.

- `perf_counter` — измерение времени выполнения алгоритмов.
- `sqrt` — вычисление квадратного корня.

In [9]:
from time import perf_counter
from math import sqrt

## 1) Полный перебор

В данном решении выполняется полный перебор всех возможных значений `a`, `b` и `c` в диапазоне от `1` до `999`.

Для каждой тройки одновременно проверяются два условия:

- выполняется теорема Пифагора;
- сумма чисел равна `1000`.

Если оба условия выполняются, возвращается произведение `abc`.

### Алгоритм

1. Последовательно перебираются все возможные значения `a`.
2. Для каждого `a` перебираются все значения `b`.
3. Для каждой пары `(a, b)` перебираются все значения `c`.
4. Проверяется выполнение условий:
   - `a² + b² = c²`;
   - `a + b + c = 1000`.
5. При нахождении подходящей тройки возвращается произведение `abc`.

### Особенности

- Используется полный перебор всех возможных троек.
- Проверяются все комбинации чисел независимо от их порядка.
- Реализация максимально проста и наглядна.
- Временная сложность составляет `O(n³)`.
- Большая часть проверок оказывается лишней, поэтому алгоритм работает медленно.

### Вывод

Полный перебор позволяет решить задачу без использования дополнительных математических свойств. Однако из-за большого количества проверяемых комбинаций данный подход является наименее эффективным.

In [10]:
def get_result_1():
    for a in range(1, 1000):
        for b in range(1, 1000):
            for c in range(1, 1000):
                if a**2 + b**2 == c**2 and a + b + c == 1000:
                    return a * b * c
    return None
                

start = perf_counter()
print(get_result_1())
end = perf_counter()

print(f'{end - start:.5f}')

31875000
22.87403


## 2) Полный перебор с учётом условия `a < b < c`

Во втором решении используется тот же полный перебор, однако учитывается одно из условий задачи:

$$
a < b < c
$$

Благодаря этому перебираются только те комбинации, которые потенциально могут образовывать тройку Пифагора. Это позволяет избежать большого количества лишних проверок.

### Алгоритм

1. Последовательно перебираются значения `a`.
2. Для каждого `a` значения `b` начинаются с `a + 1`.
3. Для каждой пары `(a, b)` значения `c` начинаются с `b + 1`.
4. Проверяется выполнение условий:
   - `a² + b² = c²`;
   - `a + b + c = 1000`.
5. При нахождении подходящей тройки возвращается произведение `abc`.

### Особенности

- Используется полный перебор с учётом условия `a < b < c`.
- Исключаются повторяющиеся и заведомо неподходящие комбинации.
- Количество проверок значительно уменьшается по сравнению с первым решением.
- Общая временная сложность остаётся `O(n³)`.
- Решение заметно быстрее полного перебора без ограничений.

### Вывод

Использование условия `a < b < c` позволяет существенно сократить число рассматриваемых комбинаций, не изменяя основную идею алгоритма. Несмотря на это, метод полного перебора по-прежнему остаётся недостаточно эффективным для больших диапазонов.

In [11]:
def get_result_2():
    for a in range(1, 1000 - 2):
        for b in range(a + 1, 1000 - 1):
            for c in range(b + 1, 1000):
                if a**2 + b**2 == c**2 and a + b + c == 1000:
                    return a * b * c
    return None
                
                    
start = perf_counter()
print(get_result_2())
end = perf_counter()

print(f'{end - start:.5f}')

31875000
10.23000


## 3) Вычисление третьего элемента тройки

В данном решении значение `c` больше не перебирается отдельно.

Для каждой пары чисел `a` и `b` третье число вычисляется по теореме Пифагора:

$$
c = \sqrt{a^2 + b^2}
$$

После этого проверяется, является ли полученное значение целым числом и выполняется ли условие

$$
a + b + c = 1000
$$

### Алгоритм

1. Последовательно перебираются значения `a`.
2. Для каждого `a` перебираются значения `b`, начиная с `a + 1`.
3. Значение `c` вычисляется по формуле:
   ```python
   c = sqrt(a**2 + b**2)
   ```
4. Проверяется, является ли `c` целым числом.
5. Проверяется условие `a + b + c == 1000`.
6. При нахождении подходящей тройки возвращается произведение `abc`.

### Особенности

- Отдельный цикл по `c` больше не используется.
- Третье число вычисляется напрямую по теореме Пифагора.
- Количество проверяемых комбинаций значительно уменьшается.
- Временная сложность снижается с `O(n³)` до `O(n²)`.
- Для вычисления квадратного корня используется функция `sqrt()`.

### Вывод

Вычисление `c` по значениям `a` и `b` позволяет отказаться от третьего вложенного цикла. Благодаря этому решение становится значительно эффективнее предыдущих вариантов, сохраняя при этом простую и понятную логику.

In [12]:
def get_result_3():
    for a in range(1, 1000 - 2):
        for b in range(a + 1, 1000 - 1):
            c = sqrt(a**2 + b**2)
            if c % 1 == 0 and a + b + c == 1000:
                return a * b * int(c)
    return None
                
                    
start = perf_counter()
print(get_result_3())
end = perf_counter()

print(f'{end - start:.5f}')

31875000
0.05357


## 4) Ограничение диапазона поиска

В данном решении используется та же идея вычисления третьего элемента тройки по теореме Пифагора, однако дополнительно уменьшается диапазон перебора значений `a` и `b`.

Из условия

$$
a < b < c
$$

и равенства

$$
a + b + c = 1000
$$

следует:

- число `a` не может быть больше `1000 / 3`;
- число `b` не может быть больше `1000 / 2`.

Благодаря этому количество рассматриваемых пар значительно сокращается.

### Алгоритм

1. Перебираются значения `a` от `1` до `333`.
2. Для каждого `a` перебираются значения `b` от `a + 1` до `500`.
3. Значение `c` вычисляется по формуле:
   ```python
   c = sqrt(a**2 + b**2)
   ```
4. Проверяется, является ли `c` целым числом.
5. Проверяется условие `a + b + c == 1000`.
6. При нахождении подходящей тройки возвращается произведение `abc`.

### Особенности

- Используется теорема Пифагора для вычисления `c`.
- Диапазоны перебора уменьшаются за счёт анализа условия задачи.
- Количество проверяемых комбинаций становится значительно меньше.
- Временная сложность остаётся `O(n²)`, однако уменьшается константа.

### Вывод

Использование математических ограничений позволяет сократить пространство поиска без изменения идеи алгоритма. Благодаря уменьшению диапазонов перебора данное решение выполняется быстрее предыдущих.

In [13]:
def get_result_4():
    for a in range(1, 333 + 1):
        for b in range(a + 1, 500 + 1):
            c = sqrt(a**2 + b**2)
            if c % 1 == 0 and a + b + c == 1000:
                return a * b * int(c)
    return None
                
                    
start = perf_counter()
print(get_result_4())
end = perf_counter()

print(f'{end - start:.5f}')

31875000
0.01984


## 5) Вычисление второго элемента по формуле

В данном решении перебирается только значение `a`.

Значение `b` вычисляется напрямую из системы условий:

$$
a^2 + b^2 = c^2
$$

$$
a + b + c = 1000
$$

Из второго равенства:

$$
c = 1000 - a - b
$$

Подставим это значение в теорему Пифагора:

$$
a^2 + b^2 = (1000 - a - b)^2
$$

После раскрытия скобок и упрощения получаем:

$$
b = \frac{500000 - 1000a}{1000 - a}
$$

Именно эта формула используется в программе:

```python
b = (500_000 - 1_000 * a) / (1_000 - a)
```

### Алгоритм

1. Перебираются возможные значения `a` от `1` до `333`.
2. Для каждого `a` по формуле вычисляется значение `b`.
3. Проверяется, является ли `b` целым числом.
4. Значение `c` вычисляется по теореме Пифагора.
5. Проверяется, является ли `c` целым числом.
6. При нахождении подходящей тройки возвращается произведение `abc`.

### Особенности

- Перебирается только одна переменная.
- Значение `b` вычисляется напрямую по формуле.
- Значение `c` определяется по теореме Пифагора.
- Временная сложность снижается до `O(n)`.
- Количество проверок значительно меньше, чем в предыдущих решениях.

### Вывод

Преобразование исходных условий позволяет отказаться от двойного перебора и вычислять `b` напрямую по известному значению `a`. Благодаря этому решение становится значительно быстрее и использует только один цикл.

In [14]:
def get_result_5():
    for a in range(1, 333 + 1):
        b = (500_000 - 1_000 * a) / (1_000 - a)
        if b % 1 == 0:
            c = sqrt(a**2 + b**2)
            if c % 1 == 0:
                return a * int(b) * int(c)
    return None


start = perf_counter()
print(get_result_5())
end = perf_counter()

print(f'{end - start:.5f}')

31875000
0.00014


## Сравнение времени выполнения

Для сравнения производительности используется функция `timeit()`, предназначенная специально для измерения времени выполнения небольших участков кода.

Каждый алгоритм запускается **10 раз**, после чего общее время делится на количество запусков. Таким образом получается **среднее время выполнения одного запуска**.

In [15]:
from timeit import timeit

funcs = (
    get_result_1,
    get_result_2,
    get_result_3,
    get_result_4,
    get_result_5,
)

for i, func in enumerate(funcs, start=1):
    tm = timeit(lambda: func(), number=10) / 10
    print(f'Алгоритм {i}: {tm:.5f}')

Алгоритм 1: 23.21680
Алгоритм 2: 9.45827
Алгоритм 3: 0.04341
Алгоритм 4: 0.01912
Алгоритм 5: 0.00005


## Заключение

В ходе решения задачи были рассмотрены пять различных подходов к поиску особой тройки Пифагора.

Первый алгоритм основан на полном переборе всех возможных значений `a`, `b` и `c`. Несмотря на простоту реализации, он выполняет огромное количество лишних проверок и оказался самым медленным.

Во втором решении используется дополнительное условие `a < b < c`, благодаря чему исключаются повторяющиеся комбинации. Это позволило почти вдвое сократить время выполнения, однако временная сложность алгоритма осталась кубической.

Третий подход значительно эффективнее: значение `c` вычисляется напрямую по теореме Пифагора, что позволяет отказаться от третьего вложенного цикла и снизить сложность до `O(n²)`.

В четвёртом решении дополнительно используются ограничения, вытекающие из условия задачи. Сокращение диапазонов перебора уменьшает количество проверок и ещё больше ускоряет поиск решения.

Наиболее эффективным оказалось пятое решение. После преобразования системы уравнений значение `b` вычисляется непосредственно по формуле, поэтому перебирается только одна переменная. Такой подход снижает сложность алгоритма до `O(n)` и обеспечивает максимальную производительность.

Среднее время выполнения алгоритмов:

- **Алгоритм 1:** `23.21680` с
- **Алгоритм 2:** `9.45827` с
- **Алгоритм 3:** `0.04341` с
- **Алгоритм 4:** `0.01912` с
- **Алгоритм 5:** `0.00005` с

Полученные результаты наглядно показывают, насколько большое влияние оказывает выбор алгоритма. Переход от полного перебора к использованию математических свойств задачи позволил сократить время выполнения более чем в **460 000 раз**. Это отличный пример того, что грамотный анализ условия зачастую приносит значительно больший выигрыш, чем любые низкоуровневые оптимизации кода.